x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,125 @@
+import org.elasticsearch.xpack.core.enrollment.CreateEnrollmentTokenAction;
+import org.elasticsearch.xpack.core.enrollment.CreateEnrollmentTokenRequest;
+import org.elasticsearch.xpack.core.enrollment.CreateEnrollmentTokenResponse;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.ssl.CertParsingUtils;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authc.support.ApiKeyGenerator;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * Transport action responsible for creating an enrollment token based on a request.
+ */
+
+public class TransportCreateEnrollmentTokenAction
+    extends HandledTransportAction<CreateEnrollmentTokenRequest, CreateEnrollmentTokenResponse> {
+
+    public static final long ENROLL_API_KEY_EXPIRATION_SEC = 30*60;
+
+    private final ApiKeyGenerator generator;
+    private final SecurityContext securityContext;
+    private final Environment environment;
+    private final NodeService nodeService;
+
+    @Inject
+    public TransportCreateEnrollmentTokenAction(TransportService transportService, ActionFilters actionFilters, ApiKeyService apiKeyService,
+                                                SecurityContext context, CompositeRolesStore rolesStore,
+                                                NamedXContentRegistry xContentRegistry, Environment environment, NodeService nodeService) {
+        super(CreateEnrollmentTokenAction.NAME, transportService, actionFilters, CreateEnrollmentTokenRequest::new);
+        this.generator = new ApiKeyGenerator(apiKeyService, rolesStore, xContentRegistry);
+        this.securityContext = context;
+        this.environment = environment;
+        this.nodeService = nodeService;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateEnrollmentTokenRequest request,
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        createEnrolmentToken(listener);
+    }
+
+    private void createEnrolmentToken(ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            final TimeValue expiration = TimeValue.timeValueSeconds(ENROLL_API_KEY_EXPIRATION_SEC);
+            final List<RoleDescriptor> roleDescriptors = new ArrayList<>(1);
+            final String[] clusterPrivileges = { "enroll" };
+            final RoleDescriptor roleDescriptor = new RoleDescriptor("create_enrollment_token", clusterPrivileges, null, null);
+            roleDescriptors.add(roleDescriptor);
+            CreateApiKeyRequest apiRequest = new CreateApiKeyRequest("enrollment_token_API_key_" + UUIDs.base64UUID(),
```

Comment:
we do we prepend the docID with `"enrollment_token_API_key_` ?

Reply:
Removed